h/realtime.py

```diff
@@ -64,29 +70,45 @@ class Publisher:
     """
 
     def __init__(self, request):
-        self.connection = get_connection(request.registry.settings)
+        self.connection = get_connection(request.registry.settings, fail_fast=True)
         self.exchange = get_exchange()
 
     def publish_annotation(self, payload):
-        """Publish an annotation message with the routing key 'annotation'."""
+        """Publish an annotation message with the routing key 'annotation'.
+
+        :raise RealtimeMessageQueueError: When we cannot queue the message
+        """
         self._publish("annotation", payload)
 
     def publish_user(self, payload):
-        """Publish a user message with the routing key 'user'."""
+        """Publish a user message with the routing key 'user'.
+
+        :raise RealtimeMessageQueueError: When we cannot queue the message
+        """
         self._publish("user", payload)
 
     def _publish(self, routing_key, payload):
         retry_policy = {"max_retries": 5, "interval_start": 0.2, "interval_step": 0.3}
 
-        with producer_pool[self.connection].acquire(block=True) as producer:
-            producer.publish(
-                payload,
-                exchange=self.exchange,
-                declare=[self.exchange],
-                routing_key=routing_key,
-                retry=True,
-                retry_policy=retry_policy,
-            )
+        try:
+            with producer_pool[self.connection].acquire(
+                block=True, timeout=1
+            ) as producer:
+                producer.publish(
+                    payload,
+                    exchange=self.exchange,
+                    declare=[self.exchange],
+                    routing_key=routing_key,
+                    retry=True,
+                    retry_policy=retry_policy,
+                )
+
+        except (OperationalError, LimitExceeded) as err:
+            # If we fail to connect (OperationalError), or we don't get a
+            # producer from the pool in time (LimitExceeded) raise
+            LOG.error("Failed to queue realtime message with error %s", err)
+            LOG.debug("Failed message payload was: %s", payload)
+            raise RealtimeMessageQueueError() from err
```

Comment:
I think this logging may be superfluous as the raised exception should appear in Sentry, replete with all kinds of useful debugging info that you wouldn't find in Papertrail, and we could set up alarms etc. I think you'd have to set up Sentry locally to test that though